src/linear_algebra/bilinear_form.lean

```diff
@@ -1016,3 +1036,322 @@ begin
+{ carrier := { m | ∀ n ∈ N, is_ortho B m n },
+  zero_mem' := λ x _, is_ortho_zero_left x,
+  add_mem' := λ x y hx hy n hn,
+    by rw [is_ortho, add_left, show B x n = 0, by exact hx n hn,
+        show B y n = 0, by exact hy n hn, zero_add],
+  smul_mem' := λ c x hx n hn,
+    by rw [is_ortho, smul_left, show B x n = 0, by exact hx n hn, mul_zero] }
+
+variables {N L : submodule R M}
+
+@[simp] lemma mem_orthogonal_iff {N : submodule R M} {m : M} :
+  m ∈ B.orthogonal N ↔ ∀ n ∈ N, is_ortho B m n := iff.rfl
+
+lemma orthogonal_le (h : N ≤ L) : B.orthogonal L ≤ B.orthogonal N :=
+λ _ hn l hl, hn l (h hl)
+
+lemma le_orthogonal_orthogonal (hB : refl_bilin_form.is_refl B) :
+  N ≤ B.orthogonal (B.orthogonal N) := λ n hn m hm, hB _ _ (hm n hn)
+
+end orthogonal
+
+/-- A set of vectors `v` is orthogonal with respect to some bilinear form `B` if and only if for
+  all `i ≠ j`, `B (v i) (v j) = 0`. -/
+def is_ortho' {n : Type w} (B : bilin_form R M) (v : n → M) : Prop :=
+∀ i j : n, i ≠ j → B.is_ortho (v j) (v i)
+
+lemma is_ortho'_def {n : Type w} {B : bilin_form R M} {v : n → M} :
+  B.is_ortho' v ↔ ∀ i j : n, i ≠ j → B (v j) (v i) = 0 := iff.rfl
+
+/-- The restriction of a bilinear form on a submodule. -/
+@[simps apply]
+def restrict (B : bilin_form R M) (W : submodule R M) : bilin_form R W :=
+{ bilin := λ a b, B a b,
+  bilin_add_left := λ _ _ _, add_left _ _ _,
+  bilin_smul_left := λ _ _ _, smul_left _ _ _,
+  bilin_add_right := λ _ _ _, add_right _ _ _,
+  bilin_smul_right := λ _ _ _, smul_right _ _ _}
+
+lemma restrict_sym (B : bilin_form R M) (hB : sym_bilin_form.is_sym B)
+  (W : submodule R M) : sym_bilin_form.is_sym $ B.restrict W :=
+λ x y, hB x y
+
+/-- A nondegenerate bilinear form is a bilinear form such that the only element that is orthogonal
+  to every other element is `0`. -/
+def nondegenerate (B : bilin_form R M) : Prop :=
+  ∀ m : M, (∀ n : M, B m n = 0) → m = 0
+
+variables {n : Type w} [fintype n]
+
+/-- A bilinear form is nondegenerate if and only if it has a trivial kernel. -/
+theorem nondegenerate_iff_ker_eq_bot {B : bilin_form R₂ M₂} :
+  B.nondegenerate ↔ B.to_lin.ker = ⊥ :=
+begin
+  rw linear_map.ker_eq_bot',
+  split; intro h,
+  { refine λ m hm, h _ (λ x, _),
+    rw [← to_linear_map_apply, hm], refl },
+  { intros m hm, apply h,
+    ext, exact hm x }
+end
+
+/-- The linear map induced by a nondegenerate bilinear form is injective. -/
+lemma to_lin_injective {B : bilin_form R₃ M₃} (hB : B.nondegenerate) :
+  function.injective B.to_lin :=
+B.to_lin.to_add_monoid_hom.injective_iff.2 (λ a ha, hB _ (linear_map.congr_fun ha))
+
+/-- Let `B` be a symmetric, nondegenerate bilinear form on a nontrivial module `M` over the ring
+  `R` with invertible `2`. Then, there exists some `x : M` such that `B x x ≠ 0`. -/
+lemma exists_bilin_form_self_neq_zero [htwo : invertible (2 : R)] [nontrivial M]
+  {B : bilin_form R M} (hB₁ : B.nondegenerate) (hB₂ : sym_bilin_form.is_sym B) :
+  ∃ x, ¬ B.is_ortho x x :=
+begin
```

Comment:
I'm not sure if I can use anything from `linear_algebra.quadratic_form` since that file imports `linear_algera.bilinear_form`. I can perhaps make a new file but I feel like the final result should belong in `bilinear_form`